core/templates/dev/head/pages/exploration-editor-page/settings-tab/settings-tab.directive.ts

```diff
@@ -118,6 +118,324 @@ angular.module('oppia').directive('settingsTab', [
             CATEGORIES_TO_COLORS, DEFAULT_CATEGORY_ICON, DEFAULT_COLOR,
             EXPLORATION_TITLE_INPUT_FOCUS_LABEL, TAG_REGEX) {
           var ctrl = this;
+
+          var CREATOR_DASHBOARD_PAGE_URL = '/creator_dashboard';
+          var EXPLORE_PAGE_PREFIX = '/explore/';
+
+          ctrl.getExplorePageUrl = function() {
+            return (
+              window.location.protocol + '//' + window.location.host +
+              EXPLORE_PAGE_PREFIX + ctrl.explorationId);
+          };
+
+          ctrl.refreshSettingsTab = function() {
+            // Ensure that ExplorationStatesService has been initialized
+            // before getting the state names from it. Otherwise, navigating
+            // to the settings tab directly (by entering a URL that ends with
+            // /settings) results in a console error.
+
+            ctrl.hasPageLoaded = false;
+            ExplorationDataService.getData().then(function() {
+              if (ExplorationStatesService.isInitialized()) {
+                var categoryIsInSelect2 = ctrl.CATEGORY_LIST_FOR_SELECT2.some(
+                  function(categoryItem) {
+                    return (
+                      categoryItem.id ===
+                          ExplorationCategoryService.savedMemento);
+                  }
+                );
+                // If the current category is not in the dropdown, add it
+                // as the first option.
+                if (!categoryIsInSelect2 &&
+                    ExplorationCategoryService.savedMemento) {
+                  ctrl.CATEGORY_LIST_FOR_SELECT2.unshift({
+                    id: ExplorationCategoryService.savedMemento,
+                    text: ExplorationCategoryService.savedMemento
+                  });
+                }
+
+                ctrl.stateNames = ExplorationStatesService.getStateNames();
+              }
+              ctrl.hasPageLoaded = true;
+            });
+          };
+
+          $scope.$on('refreshSettingsTab', ctrl.refreshSettingsTab);
```

Comment:
This and other $scope.$on should be in $onInit (you are binding a listener).

Make sure to preserve the order with regards to the original file.